src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -70,40 +80,67 @@
                     if (incoming.getCount() != lastPoint) {
                         lastPoint = incoming.getCount();
                         int hide = incoming.getCount() - outgoing.getCount() - adjustment;
-                        LOGGER.finest("hiding " + hide + " @" + (outgoing.getCount() + adjustment));
-                        text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
-                        adjustment += hide;
+                        // If openTags is not empty, but there are no escape sequences directly on this line, or if we
+                        // are just closing tags at the end of the line, there is nothing to hide.
+                        if (hide != 0) {
+                            LOGGER.log(Level.FINEST, "hiding {0} @{1}", new Object[] { hide, outgoing.getCount() + adjustment });
+                            text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
+                            adjustment += hide;
+                        }
                     }
                 }
+                public void emitHtmlDirect(String html) {
+                    text.addMarkup(incoming.getCount(), html);
+                }
             }
             EmitterImpl emitter = new EmitterImpl();
-            CountingOutputStream incoming = new CountingOutputStream(new AnsiHtmlOutputStream(outgoing, colorMap, emitter));
+            AnsiHtmlOutputStream ansiOs = new AnsiHtmlOutputStream(outgoing, colorMap, emitter);
+            CountingOutputStream incoming = new CountingOutputStream(ansiOs);
             emitter.incoming = incoming;
             try {
+                for (AnsiAttributeElement element : openTags) {
+                    // We need to reopen tags that were still open at the end of the previous line in the
+                    // so the stream's state is correct in case those tags are closed mid-line.
+                    ansiOs.openTag(element);
+                }
                 byte[] data = s.getBytes(charset);
                 for (int i = 0; i < data.length; i++) {
                     // Do not use write(byte[]) as offsets in incoming would not be accurate.
                     incoming.write(data[i]);
                 }
+                nextOpenTags = ansiOs.getOpenTags();
+                // Close any tags that are still open at the end of the line.
+                ansiOs.closeOpenTags(null);
             } catch (IOException x) {
                 LOGGER.log(Level.WARNING, null, x);
             }
             LOGGER.finer(() -> "\"" + StringEscapeUtils.escapeJava(s) + "\" → \"" + StringEscapeUtils.escapeJava(text.toString(true)) + "\"");
         }
-        return this;
+        return openTags == nextOpenTags
+                ? this
+                : new ColorConsoleAnnotator(colorMapName, charset, nextOpenTags);
```

Comment:
Based on my reading of the `ConsoleAnnotator` Javadocs, if we want to track state from line-to-line then we need to handle it by returning a new instance like this. Maybe wrong to `return this` instead of creating a new state when the open tags match.

Reply:
I thought the suggested fix for JENKINS-34019 was just to close all outstanding tags at the end of a line, as if this were a user error, not to try to make them apply across lines.

Reply:
> When a style continues to the next line, the ConsoleAnnotatorFactory can insert a closing tag at the end of the line and an opening tag at the beginning of the next line. This prevents the timestamps from being affected.

[Full comment](https://issues.jenkins-ci.org/browse/JENKINS-34019?focusedCommentId=272506&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-272506). What you propose would certainly be simpler, but I'm not sure what the expected behavior is. What do you think should be the behavior for [the test case I added](https://github.com/jenkinsci/ansicolor-plugin/blob/a95ea81c6d529e16da44437d1cf12b025390a4e2/src/test/java/hudson/plugins/ansicolor/AnsiColorBuildWrapperTest.java#L99)?

Reply:
I feel like it does not matter much, since it is unlikely people would have multiline formatting intentionally.

Reply:
#145 shows a problem with output of `terraform` at the command line, and based on a screenshot of the text it looks like it might be using multiline formatting, but I'm not sure. I'll leave this as-is for now until I have time to investigate in more depth.

Reply:
Confirmed via

```sh
docker run --rm hashicorp/terraform plan 2>&1 | od -a
```

(Though it also looks like the Terraform output is buggy—ANSI sequences are not nestable, and it is emitting several `esc   [   0   m` sequences in a row.)